Remove use of finalizers in JNA and improve concurrency for `Memory`, `CallbackReference` and `NativeLibrary`

matthiasblaesing opened the pull request.

matthiasblaesing commented:
The finalizer mechanism is deprecated for removal: https://mail.openjdk.java.net/pipermail/jdk-dev/2021-November/006279.html
So while #finalize was warned about for a long time, now removal must happen.

dbwiddis commented:
FYI, I've got a report of running out of memory that I'm pretty sure is related to excessive use of finalizer in the `Memory` allocations underlying `HANDLEByReference` and `Structure` classes when reading process and thread performance counters.  I'll copy over the new Memory code and attempt to test the code and quantify the improvement, but I personally am now a lot more interested in this PR than I was before.

matthiasblaesing commented:
@dbwiddis - I reread the changes and the API is still intact. There is a potential behavioral change, but I think it should not be common that people overrode `#dispose` and expected it to be called from `#finalize`. It won't get better, so it would be nice if you could give it another look.

dbwiddis commented on src/com/sun/jna/Cleaner.java:
```diff
@@ -0,0 +1,152 @@
+                        if(ref instanceof CleanerRef) {
+                            ((CleanerRef) ref).clean();
+                        }
+                    } catch (InterruptedException ex) {
```
Should we set `Thread.currentThread.interrupt()` here?  I recognize that the thread will essentially terminate on an interrupt and it probably doesn't matter in practice, but will keep static analysis tools happier.

matthiasblaesing replied:
I don't get the intention. InterruptedException is raised if the waiting thread is interrupted (in this case the cleaner thread). Why would an analyser expect that the threads interrupts itself again?

dbwiddis replied:
Analyzers don't like when you catch an interrupted exception and do nothing.  The general guidance is to either re-throw the exception or re-set the interrupted flag so that any upstream code recognizes the interruption.

As I said it probably doesn't matter in this case as the thread just terminates. (I'm not sure if it would terminate with a different exit code with that flag or not, or if that matters.)

This article says what I'm saying a lot better than I can:  https://programming.guide/java/handling-interrupted-exceptions.html

dbwiddis commented on src/com/sun/jna/Cleaner.java:
```diff
@@ -0,0 +1,152 @@
+        return ref;
+    }
+
+    private synchronized boolean remove(CleanerRef ref) {
```
In theory this could be moved inside the `CleanerRef` class, but I like it out here with the `add()` and `register()` methods.

dbwiddis commented on src/com/sun/jna/Cleaner.java:
```diff
@@ -0,0 +1,152 @@
+        }
+    }
+
+    public static interface Cleanable {
```
Since this uses the same name as a JDK9+ class, recommend a javadoc indicating that it's intended to serve the same purpose.  (I assume you based your code on the JDK9 Cleaner?)

matthiasblaesing replied:
Yes this is modelled after the JDK9+ Cleaner. The cleaner is intended to be used by JNA internal code. I made this cleaner in the update by moving the cleaner into the "internal" package and adjust the documentation slightly.

dbwiddis commented on src/com/sun/jna/Memory.java:
```diff
@@ -40,91 +44,15 @@
-            }
-        }
-    }
+public class Memory extends Pointer implements Closeable {
```
`AutoCloseable` would really be preferred but that would require JDK7.  

I'm wondering if it would be good to also make some of the key JNA classes that hold an internal `Memory` object `Closeable` as well, having their `close()` method invoke the memory's `close()`.  I'm thinking of `Structure` and `ByReference` in particular.

I don't think it would break compatibility but will let downstream projects use Structure and FooByReference allocations in try-with-resources blocks.

matthiasblaesing replied:
I'm inclined to say yes, but IMHO this is out of scope for this change.

dbwiddis replied:
Agreed. Once this is merged I might put together a new PR to add this.

dbwiddis commented on src/com/sun/jna/Memory.java:
```diff
@@ -137,71 +65,17 @@ public static void purge() {
-                entry = entry.prev;
-                index--;
+        // use a copy since dispose() modifies the map
+        Collection<Reference<Memory>> refs = new LinkedList<Reference<Memory>>(allocatedMemory.values());
```
Since this is just a temporary variable is there any value in using `LinkedList` vs. `ArrayList`?

matthiasblaesing replied:
Valid point pushed updated version.

dbwiddis reviewed (approved):
LGTM, but a few comments/suggestions.

matthiasblaesing commented:
Ok - this needs more work - the unittests became flaky. I first thought, that the cleaner implementation of `ProxyObject` was broken, but that does not explain unrelated failures. I suspect, that the `Memory` objects get quicker collected and thus we see race effects.

matthiasblaesing commented:
This is the problematic part:

```
    [junit] Exception in thread "JNA Cleaner" java.lang.Error: Invalid memory access
    [junit] 	at com.sun.jna.Native._getPointer(Native Method)
    [junit] 	at com.sun.jna.Native.getPointer(Native.java:2263)
    [junit] 	at com.sun.jna.Pointer.getPointer(Pointer.java:642)
    [junit] 	at com.sun.jna.platform.win32.COM.COMInvoker._invokeNativeInt(COMInvoker.java:37)
    [junit] 	at com.sun.jna.platform.win32.COM.Unknown.Release(Unknown.java:80)
    [junit] 	at com.sun.jna.platform.win32.COM.util.ProxyObject$Disposer.run(ProxyObject.java:747)
    [junit] 	at com.sun.jna.internal.Cleaner$CleanerRef.clean(Cleaner.java:127)
    [junit] 	at com.sun.jna.internal.Cleaner$1.run(Cleaner.java:60)
    [junit] Testsuite: com.sun.jna.platform.win32.COM.util.HybdridCOMInvocationTest
```

I still try to understand how we could end there.

dbwiddis commented:
> ```
>     [junit] 	at com.sun.jna.platform.win32.COM.util.ProxyObject$Disposer.run(ProxyObject.java:747)
> ```

I suspect a race condition in `ObjectFactory.disposeAll()` (called in `testOfficeInvocationProblemCOMUtil()` in the problem test suites) which calls `ProxyObject.dispose()` (which no longer synchronizes on anything), vs. `ProxyObject.Disposer.dispose()`.

dbwiddis commented:
FYI, long-running program with the Memory portion of this PR resolved the out-of-memory issues seen earlier with too many finalizers.  So if we can resolve the test failures this looks like a good improvement.

matthiasblaesing commented:
So what happened: I removed the changes to the platform code, added stabilization to some of the regularly failing tests to get noise down or get better diagnostics. Multiple runs got through cleanly or reported errors that are "consistent" with breaks, that can happen in a shared system.

I intent to merge this at the end of the week end.
